
<!-- saved from url=(0068)http://bozeman.mbt.washington.edu/compbio/mbt599/homework1_2018.html -->
<html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>Genome 540 Homework Assignment 1</title>
</head>
<body bgcolor="#ffffff">
<h3 align="center">Genome 540 Homework Assignment 1</h3>
<h3 align="center">Due Sunday Jan. 14</h3>

<p><b> Late homework policy is described on course web page. </b>

</p><ol> 
<li style="margin-bottom: 10px;"> Download and begin reading <a href="http://dx.doi.org/10.1038/35057062"> <i>Initial sequencing and analysis of the human genome.</i> The Genome
International Sequencing Consortium. <i>Nature</i> 409, 860-921 (15
February 2001) </a>. (To print this out, I would recommend the
pdf format which corresponds exactly to the printed version, rather
than the html format.) For next week, read:
<ul> 
<li style="margin-bottom: 5px;"> Introduction and background (pp. 860-863, up to but not including "Strategic issues")
</li><li style="margin-bottom: 5px;"> The section "Broad genomic landscape" (pp. 875-879, up to but not including "Repeat content of the human genome")
</li><li style="margin-bottom: 5px;"> The section "Gene content of the human genome" (starting p.
892) up to but not including "Comparative proteome analysis" (p. 901).
</li></ul>

</li>
<li style="margin-bottom: 10px;"> Write a program that implements the method described in Lecture 2
to find the longest exactly matching subsequence between two
sequences.  Specifically, your program should:
<ul>
<li style="margin-bottom: 5px;">Read two input files in
"FASTA" format (i.e. having a header line which starts with the
character "&gt;" and includes the sequence name, with the sequence
itself following on subsequent lines), each of which contains one of
the sequences to be compared. As a data check, your program should
count the number of bases of each type (i.e. A, C, T, G, or N) in each sequence.</li>
<li style="margin-bottom: 5px;">Use the described algorithm to find the longest subsequence
present in both sequences. If there are several different perfectly
repeated subsequences of the same length, find all of them. When you
are looking for these, consider <i>both strands</i> of each sequence
(i.e. both the sequence given in the FASTA file, which is sometimes
called the 'forward' strand, and its reverse complement, which is
sometimes called the 'reverse' strand) simultaneously, so that if the
repeated subsequence happens to occur on different strands in the two
sequences you will still find it.</li>
<li style="margin-bottom: 5px;">The easiest way to do this is to
create and store in memory a single sequence of length 2N + 2M + 4
(where the two input sequences have lengths N and M bases
respectively) constructed by concatenating together the two sequences
and their reverse complements (each terminated by a null character to
terminate the strings, so that the lexicographic sort does not read
across the sequence boundaries), and then create a single list of
pointers to each position in this merged sequence, which you then
sort. If a longest subsequence is present multiple times in either 
genome, please report <i>all locations</i> of the subsequence.</li>
</ul>

</li><li style="margin-bottom: 10px;"> Using this program, you should then find the longest exactly matching sequences in the 
<a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Acetobacterium_woodii_DSM_1030_uid60713/CP002987.fna"> Acetobacterium woodii genome</a> and the 
<a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Cyanobacterium_PCC_10605_uid158691/CP003947.fna"> Cyanobacterium aponinum genome</a>.
Once you find the longest match(es) using your program, you should then try to figure out what biological feature they correspond to. 
You can do this by exploring the features tables for both organisms 
<a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Acetobacterium_woodii_DSM_1030_uid60713/CP002987.gbk">here</a> 
and <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Cyanobacterium_PCC_10605_uid158691/CP003947.gbk">here</a>.

</li>
<li style="margin-bottom: 10px;"> To test whether your program is working correctly, run it first
on the test example indicated below (with two different bacterial
genomes) to see whether you get the right answer. The FASTA files for
the test examples can be found <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Mycoplasma_pneumoniae_M129_B7_uid176548/CP003913.fna">here</a> 
and <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Mycoplasma_gallisepticum_R_high__uid43299/CP001872.fna">here</a>.
The Genbank files are <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Mycoplasma_pneumoniae_M129_B7_uid176548/CP003913.gbk">here</a> 
and <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Mycoplasma_gallisepticum_R_high__uid43299/CP001872.gbk">here</a>.
In general, FASTA files for these and other bacterial genomes can be found
by going to the <a href="http://www.ncbi.nlm.nih.gov/">NCBI web site</a> and following
appropriate links. (On the NCBI site, the FASTA files containing the
full genome sequences have the suffix .fna. To find the biological
features, look in the 'Genbank format' files for the organisms (which
have the suffix .gbk on the NCBI site) and find the annotated
'feature' in each genome that overlaps the matching segment you
found. You don't need to write a program to do this -- you can just
read the .gbk file on the website.)  
 </li>

<li> You must turn in your results <i>and</i> your computer program, using the template (file format) described below. <b>Please put everything into ONE file - do not send an archive of files or a tar file.</b> After creating a plain text file (NOT a word processing document file) in this format, compress it (using either Unix compress, or gzip -- if you don't have access to either of these programs, let us know), and send it as an attachment to both Phil at phg@uw.edu
and Serena at selenay@uw.edu.


</li></ol>



Between the "=============" below is an example of a homework file with the fields filled in with the correct answers for the test case (see genomes provided in #4 above). Several items are explained in detail in the notes below the template:

<pre>
=============
Assignment: GS 540 HW1
Name: Serena Liu
Email: selenay@uw.edu
Language: Java
Runtime: 8 sec

Fasta 1: CP003913.fna
&gt;gi|440453185|gb|CP003913.1| Mycoplasma pneumoniae M129-B7, complete genome
A=249201
C=162924
G=163697
T=240551
N=0
*=816373

Fasta 2: CP001872.fna
&gt;gi|284930242|gb|CP001872.1| Mycoplasma gallisepticum str. R(high), complete genome
A=349322
C=159094
G=159365
T=344246
N=0
*=1012027

Match length: 122
Number of match strings: 2

Match string: GGGGTCTTTCCGTCCCGTTGCGCCTAACGGGTGTCTTCACCCGTACCTGGATTTCACCGAGTCTATAGCCGAGACAGTCAAGAGATGGTTACACCATTCAAGCGGGACGGAATTTACCCGAC
Description: This sequence comes from the 23S ribosomal RNA gene.

Fasta: CP001872.fna
Position: 338240
Strand: reverse

Fasta: CP003913.fna
Position: 122005
Strand: reverse

Fasta: CP001872.fna
Position: 82469
Strand: reverse

Match string: GTCGGGTAAATTCCGTCCCGCTTGAATGGTGTAACCATCTCTTGACTGTCTCGGCTATAGACTCGGTGAAATCCAGGTACGGGTGAAGACACCCGTTAGGCGCAACGGGACGGAAAGACCCC
Description: This sequence comes from the 23S ribosomal RNA gene.

Fasta: CP001872.fna
Position: 82469
Strand: forward

Fasta: CP003913.fna
Position: 122005
Strand: forward

Fasta: CP001872.fna
Position: 338240
Strand: forward

Program:

int main() {
	do_analysis();
	return 0;
}

=============

</pre>


<h3>Details:</h3>


<p><b>Fasta:</b> put the name of the fasta file, along with the first line.


</p><p><b>Nucleotide histogram:</b> count the total number of bases ("*") and the number of times each specific base occurs.  i.e.:
</p><pre>A=349623
C=159237
G=159490
T=344450
N=0
*=1012800
</pre>


<p><b>Description:</b> Put a short identification of the shared DNA subsequence, for instance:
"This DNA sequence is the first 20 bases of an RNA polymerase gene."


</p><p><b>Position:</b> Put the location of the matching subsequence within the input sequence here.  The location should be the index of the DNA base in the sequence that is <i>closest to the beginning of the forward strand</i>. Use a coordinate system starting at 1 rather than 0.
For example, if the two chromosomal strands are:
</p><pre>			 5'-ACTGA-3'
			 3'-TGACT-5'
</pre>
and you found the subsequence TCA on the reverse strand to be the longest  match to the other sequence, then the location should be reported as 3.  If instead you found CTG on the forward strand, then the location should be reported as 2.
</body></html>