
<!-- saved from url=(0049)http://bozeman.mbt.washington.edu/compbio/mbt599/ -->
<html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>Genome 540: Computational Molecular Biology </title>
</head>
<body bgcolor="#FFFFFF">
<h2 align="center">Genome 540</h2>
<h2 align="center">Introduction to Computational Molecular Biology:</h2>

<h2 align="center"> Genome and Protein Sequence Analysis  </h2>
<h2 align="center"> (Winter Quarter 2018)  </h2>


<ul> 

<li> <b>Synopsis:</b> Together with <a href="http://noble.gs.washington.edu/~noble/genome541/">Genome 541</a>, a two-quarter
introduction to protein and DNA
sequence analysis and molecular evolution, including probabilistic models of sequences and of sequence
evolution, computational gene identification, pairwise sequence
comparison and alignment (algorithms and statistical issues), multiple
sequence alignment and evolutionary tree construction, comparative
genomics, and protein sequence/structure relationships.
  These are the central
computational methods required to determine the "periodic table of
biology", i.e. the list of proteins and their evolutionary
relationships, which can be regarded as the first stage in the growth of
molecular biology into a quantitative science.  Moreover, the
statistical and algorithmic methods used (which include maximum
likelihood estimation, hidden Markov models, dynamic programming) have
wide applicability in other areas of computational &amp; mathematical
biology.

</li><li> <b>Prerequisites:</b> Ability to write computer programs for data
analysis is essential (homework assignments will require this.) Some
prior exposure to probability and statistics, and molecular biology, is
highly desirable.

</li><li>  <b>Lectures:</b> TuTh 10:30-11:50 in <a href="http://www.washington.edu/home/maps/southcentral.html?GNOM">Foege S110</a>. We start <i>promptly</i> at 10:30.
<ul>
<li> <b>Discussion session:</b> TBA
</li></ul>
</li><li> <b>Instructors:</b> 
<ul>
<li>  <a href="http://www.phrap.org/"> <b>Phil Green</b></a> (Foege-S320B; phg (at) uw.edu)
</li><li> TA: <b>Serena Liu</b> (selenay (at) uw.edu)
</li></ul>
</li><li> <b>Office hours:</b> 
<ul>
<li> <b>Serena:</b> by appointment. 
</li><li> <b>Phil:</b> by appointment.
</li></ul>
</li><li> <b>Text:</b>
<ul>
<li> (<b>Required</b>): <i> Biological Sequence Analysis: Probabilistic Models of Proteins and Nucleic
Acids </i>
 by Richard Durbin,  S. Eddy,  A. Krogh, G. Mitchison; Cambridge University
Press, 1998. ISBN: 0521629713.  Paperback, ~$60 (available from UW Bookstore--South Campus Center Branch, or <a href="http://search.barnesandnoble.com/Biological-Sequence-Analysis/e/9780521629713/?itm=1">www.bn.com</a>, <a href="http://www.amazon.com/Biological-Sequence-Analysis-Probabilistic-Proteins/dp/0521629713/">www.amazon.com</a>).

</li><li> (<b>Required</b>): <i> Statistical Methods in Bioinformatics : An Introduction (Statistics for Biology and Health)
</i> 
 by Warren J. Ewens, Gregory R. Grant; Springer, 2005. ISBN: 0387400826. <b>NOTE THAT THIS IS THE 2d EDITION -- MAKE SURE YOU GET THIS EDITION!! </b>
 Hardbound, ~$105 (available from UW Bookstore--South Campus Center Branch, or <a href="http://search.barnesandnoble.com/Statistical-Methods-in-Bioinformatics/Warren-Ewens/e/9780387400822/?itm=1">www.bn.com</a>, <a href="http://www.amazon.com/Statistical-Methods-Bioinformatics-Introduction-Statistics/dp/0387400826/">www.amazon.com</a>).

</li></ul>
</li><li> To <b>register</b> for credit, contact Brian Giebel (bgiebel (at) u.washington.edu) to obtain an entry code. Enrollment is limited to 25 students. Auditing is allowed. 

</li><li> <b>See <a href="http://bozeman.genome.washington.edu/compbio/mbt599_2017/index.html">last year's site</a> for approximate syllabus</b> 
</li><li> <b> Grading &amp; homework policies:</b>
<ul>
<li>  The entire course grade is based on the homework assignments, which are
due weekly.
No tests or exams.
</li><li> The homework assignments involve writing programs for data analysis, and running them on a computer you have access to (we cannot provide computers). We don't require a specific language, since it is not
practical for us to grade your code, just the output from running your
programs. However it is important to use a language that allows you to
write programs that will run fast on large datasets; ideally a compiled
language such as C or C++. You may be able to get by with an interpreted
language such as Perl or Python (some people have), however there is a
definite risk that programs in these languages will take very long to
run -- which means that (i) you will need access to a very fast
computer to run them on, and/or (ii) you will need to get the program
written early enough that you can allow 2 or 3 days for the actual run and
still be able to turn in the assignment on time.
</li><li> Homework is due by 11:59 pm on the indicated date. After that it will be accepted, but penalized. Specifically, each assignment is worth 10 points, from which 1 point will be deducted for each day (or fraction thereof) that you turn it in late. The maximum deduction for being late is 3 points (even if you are more than 3 days late). If you get less than 7 points on an assignment, you are allowed to redo it and take the new score (which will be 7, i.e. 10 - 3, if there are no mistakes).
</li><li> It is OK to run your program on someone else's input datafile, and
compare outputs to see if you get the same results. However it is not OK
to share programs, or to get someone else to debug your program. A key
part of the course is being able to write and debug your own programs for
data analysis. 
</li></ul>
<p> Since important announcements (e.g. schedule changes) may be made by email in advance of lectures, <b> all attendees (whether or not registered for the course) should send their email addresses to Phil Green (phg (at) uw.edu).</b>

</p></li></ul> 

<h3> HOMEWORK ASSIGNMENTS:  </h3>
<li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/homework1_2018.html">Assignment 1</a>, due <b>Sunday Jan 14</b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/homework2_2018.html">Assignment 2</a>, due <b>Sunday Jan 21</b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/homework3_2018.html">Assignment 3</a>, due <b>Sunday Jan 28</b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/homework4_2018.html">Assignment 4</a>, due <b>Sunday Feb 4</b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/homework5_2018.html">Assignment 5</a>, due <b>Sunday Feb 11</b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/homework6_2018.html">Assignment 6</a>, due <b>Sunday Feb 18</b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/homework7_2018.html">Assignment 7</a>, due <b>Sunday Feb 25</b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/homework8_2018.html">Assignment 8</a>, due <b>Sunday Mar 4</b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/homework9_2018.html">Assignment 9</a>, due <b>Sunday Mar 11</b>



<h3> LECTURE &amp; DISCUSSION SLIDES: </h3>

<p><a href="http://bozeman.genome.washington.edu/compbio/mbt599_2001/Notation/Notation1.htm">Math Notation</a>
</p><ul>
<li> <a href="http://bozeman.genome.washington.edu/compbio/mbt599_2006/Lecture0/template.html">Biological Review </a>: Gene and genome structure in prokaryotes and eukaryotes; the genetic code &amp; codon usage; "global" genome organization. Sources and characteristics of sequence data; Genbank and other sequence databases.
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture1.pdf">Lecture 1</a>: Course overview; administrative details. Finding exact matches in sequences using suffix arrays.
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/discussions/week1.pptx"> Discussion Section 1</a>: HW1 and general programming tips. 
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture2.pdf">Lecture 2</a>: Algorithm generalities / complexity. Directed graphs, WDAGS. 
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture3.pdf">Lecture 3</a>: Dynamic programming to find highest weight path. Weighted linked lists: sequence graphs,
     WLLs for 'motif clusters'.
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/discussions/week2.pptx"> Discussion Section 2</a>: HW2, algorithms for finding minimum-weight paths, and C++ basics.
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture4.pdf">Lecture 4</a>: WLLs for numerical data; statistical issues. Finding multiple high-scoring segments. 
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture5.pdf">Lecture 5</a>: Finding multiple high-scoring segments. D-segments, relationship to 2-state HMMs. Sequence alignment &amp; evolution. 
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/discussions/week3.pptx"> Discussion Section 3</a>: D-Segment algorithm, HW3, and useful data structures.
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture6.pdf">Lecture 6</a>: Edit graphs &amp; sequence alignment. Smith-Waterman algorithm. Needleman-Wunsch algorithm. Local vs. global. Computational complexity of pairwise alignment <b>Reading: Durbin et al. section 2.1, 2.2, 2.3, 2.8, Ewens &amp; Grant 6.2, 6.4, 6.5 </b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture7.pdf">Lecture 7</a>: Multiple sequence alignment. Improved scoring of pairwise alignments: affine gap penalties, profiles. <b>Reading: Durbin et al. section 2.4, 6.1, 6.2, 6.3, 6.4 </b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/discussions/week4.pptx"> Discussion Section 4</a>: Multiple sequence alignment and edit graphs, HW4.
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture8.pdf">Lecture 8</a>: Smith-Waterman special cases. Word nucleation approaches/BLAST. Probability models on sequences. <b>Reading: Durbin et al. section 2.5 </b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture9.pdf">Lecture 9</a>: Probability models on sequences &amp; review of basic probability theory concepts. Neutralist vs selectionist interpretations. Site models.  Comparing models: likelihood ratios &amp; weight matrices. (Hypothesis testing &amp; Neyman-Pearson lemma). <b>Reading: Durbin et al. section 1.3, 11.1 Ewens &amp; Grant 1.1, 1.2, 1.12, 3.1, 3.2, 3.4, 3.6, 9.1, 9.2 </b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/discussions/week5.pptx"> Discussion Section 5</a>: Modeling translation start site motifs, HW5.
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture10.pdf">Lecture 10</a>: Review of site models, likelihood ratios, &amp; weight matrices.  (Hypothesis testing &amp; Neyman-Pearson lemma). Score distributions. Limitations of site models: gaps, failure of independence assumption. <b>Reading: Ewens &amp; Grant section 5.2, 5.3  </b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture11.pdf">Lecture 11</a>: Hidden Markov Models: Definitions, examples. Probability calculations, WDAG. Parameter estimation: Viterbi training <b>Reading: Ewens &amp; Grant section 12.1, 12.2;  Durbin et al. chapter 3 intro and sections 3.1, 3.2, 3.3 </b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/discussions/week6.pptx"> Discussion Section 6</a>: Viterbi training &amp; HW6, BLAST.
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture12.pdf">Lecture 12</a>: Review of HMM probability calculations &amp; Viterbi training. Forward &amp; forward/backward algorithms, Baum-Welch training.  <b>Reading: Ewens &amp; Grant section 12.3;  Durbin et al. 11.6, 3.4, 3.5, 3.6</b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture13.pdf">Lecture 13</a>: Baum-Welch training; techniques for finding global maxima in likelihood surface.. More HMM examples. 
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/discussions/week7.pptx"> Discussion Section 7</a>: Baum-Welch, HW7.
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture14.pdf">Lecture 14</a>: More HMM examples. Limitations of HMMs. PhyloHMMs, PhastCons. <b>Reading: Siepel et al. (link below) </b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture15.pdf">Lecture 15</a>: PhastCons. <b>Reading: Siepel et al. (link below) </b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/discussions/week8.pptx"> Discussion Section 8</a>: HW8 (identifying evolutionarily conserved segments from a multiple alignment)
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture16.pdf">Lecture 16</a>: PhastCons. 
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture17.pdf">Lecture 17</a>: PhastCons. Distribution of maximal segment scores in background sequence; Karlin-Altshul theory.
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/discussions/week9.pptx"> Discussion Section 9</a>: HW9 (D-segments with randomized data)
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Single%20cell%20genomics!.pptx">Lecture 18</a>: Single cell genomics
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/Lecture19.pdf">Lecture 19</a>: Information theory. <b>Reading: Ewens &amp; Grant section 1.14, Appendix B.10;  Durbin et al. 11.2 </b>
</li><li> <a href="http://bozeman.mbt.washington.edu/compbio/mbt599/discussions/week10.pptx"> Discussion Section 10</a>: Unix tools, version control.
</li></ul>

<h3> COURSE-RELATED PAPERS: </h3>
<ul>
<li> <a href="http://www.nature.com/nature/journal/v409/n6822/pdf/409860a0.pdf">  Nature paper on human genome sequence </a>
</li><li> <a href="http://www.nature.com/nature/journal/v420/n6915/pdf/nature01262.pdf">  Nature paper on mouse genome sequence </a>
</li><li> <a href="http://www.genome.org/cgi/content/abstract/15/8/1034">  Siepel et al. paper on PhyloHMMs &amp; sequence conservation </a>
</li><li> <a href="http://bozeman.genome.washington.edu/compbio/mbt599_2006/rabiner.pdf"> Rabiner tutorial on HMMs</a>
</li><li> <a href="http://bozeman.genome.washington.edu/compbio/mbt599_2006/hmm_scaling_revised.pdf"> HMM scaling tutorial (Tobias Mann) </a>
</li></ul>

<h3> C/C++ PROGRAMMING GUIDES: </h3>

<ul>
<li> <a href="http://www.cplusplus.com/doc/tutorial/">Tutorial at cplusplus.com</a>
</li><li> <a href="http://www.cprogramming.com/tutorial/c++-tutorial.html">Tutorial at cprogramming.com</a>
</li><li> <a href="https://www.amazon.com/How-Program-10th-Paul-Deitel/dp/0134448235">C++ <i>How to Program</i>, Deitel and Deitel</a>


</li></ul>

<h3> OTHER RELEVANT COURSES AT UW: </h3>

<ul>
<li> <a href="http://noble.gs.washington.edu/~noble/genome541/"> Genome 541</a>
</li><li> <a href="http://courses.cs.washington.edu/courses/csep590a/13sp/"> CSE P590A (Computational Biology -- Professional Masters Program)</a>
</li><li> <a href="http://www.cs.washington.edu/education/courses/527/"> CSE 527 (Computational Biology) </a>
</li><li> <a href="http://courses.washington.edu/bioinfo/"> Pathobiology 536 (Bioinformatics and Gene Sequence Analysis) </a>
</li><li>
<a href="http://evolution.gs.washington.edu/gs562/2013/">Genome 562 (Winter 2013): Population Genetics </a> 
</li><li>
<a href="http://evolution.gs.washington.edu/gs570/2012/">Genome 570 (Winter 2012): Phylogenetic Inference </a>
</li><li> <a href="http://evolution.gs.washington.edu/gs590/2007/index.html">Genome
590 (Winter 2007): Evolution and Population Genetics Seminar </a>


</li></ul>

<h3> COMPUTATIONAL BIOLOGY COURSES AT OTHER SITES: </h3>

<ul>
<li> <a href="http://www.genetics.wustl.edu/bio5495/"> Computational Molecular Biology (Washington University) </a>
</li><li> <a href="http://www.cs.tau.ac.il/~rshamir/cg/10"> Computational Genomics (Ron Shamir, Tel Aviv University) </a>
</li><li> <a href="http://biochem218.stanford.edu/">  Computational Molecular Biology (Doug Brutlag &amp; Lee Kozar, Stanford)</a>
</li><li> <a href="http://biochem228.stanford.edu/">  Computational Genomics (Doug Brutlag, Stanford)</a>
</li><li> <a href="http://stellar.mit.edu/S/course/6/fa08/6.047"> Computational Biology: Genomes, Networks, Evolution (James E. Galagan and Manolis Kellis, MIT) </a>
</li><li> <a href="http://compgen.bscb.cornell.edu/btry4840/index.php/Main_Page">Computational Genomics (Adam Siepel, Cornell)</a>
</li><li> <a href="http://www.cs.sunysb.edu/~skiena/549/">Introduction to Computational Biology (Steven Skiena, Stony Brook)</a>
</li></ul>



<h4> <a href="http://www.washington.edu/online/terms/"> UW Terms and Conditions of Use </a> </h4>

<h4> <a href="http://www.washington.edu/online/privacy/"> UW Online Privacy Statement </a> </h4>





</li></body></html>